### DISCUSSION TITLE
Persistence of input/output/context

### DISCUSSION THREAD

**Discussion by valsov** [elsa-workflows/elsa-core#1452] [Q&A] [answered]
Hello!
First of all, I really like this library, thank you for your work !

I'm a bit lost on a task linked to #1182 , I have a workflow with an input object (this one is correctly persisted and retrieved after recovery). Then, each activity will have some Inputs for the designer (those are persisted as well) and finally they will all require the previous activity output to run; except the first one that will use the workflow input.

The output is passed like so:
```cs
protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
{
    [...]
    return Done(outputForNextActivity);
}
```

This works fine as I can retrieve this output in the next activity:
```cs
protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
{
    var outputForNextActivity = context.Input;
    [...]
    return Done(outputForNextActivity);
}
```

The problem is, after a server crash, the recovered workflow instance doesn't have access to this input.
I tried looking around the documentation and the source code but with no luck. I am using Elsa 2.2.1 as I read the release notes about the data persistance.

Thank you for your time.

**Comment by valsov** [comment, marked as answer]
I'm kind of stuck and need this feature to be operational. Did someone implement activity input/output persistance ?

**Reply by sfmskywalker** [reply]
Hi @valsov , sorry for the late reply.

To make sure your activity output gets persisted (and therefore available even after a server crash, provided that the output at least got persisted), declare your output as a property. Example:

```csharp
[Output] string Output { get; set; }

protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
{
    Output = context.Input;
    [...]
    return Done();
}
```

By default, your activity output will now be persisted as part of the workflow instance. And by convention, output properties named `Output` will automatically flow as input into the next activity that is scheduled for execution.

**Reply by valsov** [reply]
Thank you for your help !
The activity output is indeed injected into the next activity Input: ```context.Input``` (where ```context``` is ```ActivityExecutionContext```).
But I'm faced with another problem: when recovering from a shutdown, the correct activity is executed but with the **Workflow** input in ```context.Input```, instead of the **previous activity** Output.
So there is a difference between normal execution and recovery execution regarding the activity Input origin.
How can I get the expected data in ```context.Input``` ?

**Reply by sfmskywalker** [reply]
Could you send me a small project that demonstrates this? Perhaps something is not right, but it would be helpful for me to see it with an example.

What you might do in any case is to give the previous activity a name (e.g. `SomeActivity1`) and then reference the output from this activity explicitly using e.g. JavaScript (e.g. `activities.SomeActivity1.Output().MaybeSomeProperty`).

**Reply by valsov** [reply]
You pointed me in the right direction, to achieve what I wanted through code only (I didn't want to do this in the designer for every single activity).
When a recovery occures I get the activity activity Input via this method:
```cs
protected ActivityInputType GetInput(ActivityExecutionContext context)
{
    var activitiesData = context.WorkflowExecutionContext.WorkflowInstance.ActivityData;
    var previousActivityData = activitiesData.ElementAt(activitiesData.Count - 2).Value;
    if (!previousActivityData.TryGetValue(nameof(Output), out var input) || input is null)
    {
        input = context.Input;
    }
    return input as ActivityInputType;
}
```

Have a nice day !
